### ISSUE TITLE
pyre can forget type narrowing in certain circumstances

### ISSUE THREAD

**Issue by GBeauregard** [facebook/pyre-check#576]
# Pyre Bug

**Bug description**
When calling functions in certain circumstances it is possible to coerce pyre to forget its type narrowing.

**Reproduction steps**
```python
import subprocess
from typing import IO

def wants_io(var: IO[str]) -> None:
    ...

def do_nothing() -> None:
    ...

with subprocess.Popen(
    "myprogram", text=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
) as result:
    assert result.stdout is not None
    do_nothing()
    reveal_type(result.stdout)
    wants_io(result.stdout)  # false positive
```

**Expected behavior**
The code should type check and `result.stdout` should stay narrowed to not include `None`. Both mypy and pyright can handle this code snippet.

**Logs**

```console
ƛ No watchman binary found.
To enable pyre incremental, you can install watchman: https://facebook.github.io/watchman/docs/install
ƛ Defaulting to non-incremental check.
ƛ Using virtual environment site-packages in search path...
ƛ Found 2 type errors!
test/min.py:15:4 Revealed type [-1]: Revealed type for `result.stdout` is `typing.Optional[IO[typing.Any]]`.
test/min.py:16:13 Incompatible parameter type [6]: In call `wants_io`, for 1st positional only parameter expected `IO[str]` but got `Optional[IO[typing.Any]]`.

```

**Additional context**
Client version: 0.9.10
Binary version: d418f5bb10fdee9648eeea3834c8ca56944bd2ce

It seems like this could be intentional? (forget assertions on function calls)

**Comment by GBeauregard** [comment]
Closing as intentional https://pyre-check.org/docs/errors/#optional-attributes
